### PR TITLE
Add isolated support to function type and anon func exper

### LINKED ISSUE #25563: Allow using `isolated` with `function-type-descriptor` and `anonymous-function-expr`
**Description:**
The proposal for `isolated` says

> A function can be declared as isolated, by using the isolated keyword before the function keyword in the following syntactic constructs:
> 
> - function-type-descriptor
> - function-defn
> - method-decl
> - method-defn
> - service-method-defn
> - anonymous-function-expr

https://github.com/ballerina-platform/ballerina-lang/pull/25423 allows this for the above except for `function-type-descriptor` and `anonymous-function-expr`.

Need to add support for these two.

IMO, it will make sense to go ahead and allow qualifierList similar to the other function constructs, and specifically allow only `isolated` for the time-being, since other qualifiers may be allowed in the future.

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$Subject

Fixes #25563

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -1017,22 +1007,39 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+
             }
         }
 
+        qualifiers = STNodeFactory.createNodeList(validatedQualifierList);
+        STNode typeDesc = STNodeFactory.createFunctionTypeDescriptorNode(qualifiers, functionKeyword,
+                funcSignature);
+
+        // Check if it is a complex type desc starting with function type
+        if (isValidTypeContinuationToken(peek())) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1038]: I think we don't need this check. `parseComplexTypeDescriptor` does the same check indirectly.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -8828,34 +8839,61 @@ private STNode parseKeyTypeConstraint(STNode keyKeywordToken) {
...
+        switch (nextToken.kind) {
+            case ISOLATED_KEYWORD:
+                qualifiers.add(consume());
+                break;
+            case FUNCTION_KEYWORD:
+                break;
+            default:
+                recover(peek(), ParserRuleContext.FUNC_TYPE_DESC_START);
+                return parseFunctionTypeDesc();
+        }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:8859]: Seems this block below is repeated in 3 places. Shall we extract it to a separate method?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -1017,22 +1007,39 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+            if (qualifier.kind == SyntaxKind.ISOLATED_KEYWORD) {
+                validatedQualifierList.add(qualifier);
+            } else if (isObjectMember && isVisibilityQualifier(qualifier)) {
                 // public or private qualifier allowed in object field.
                 visibilityQualifier = qualifier;
             } else {
-                typeDesc = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(typeDesc, qualifier,
+                functionKeyword = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(functionKeyword, qualifier,
                         DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED, ((STToken) qualifier).text());
+
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1029]: extra line?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -1017,22 +1007,39 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
         return funcDef;
     }
 
-    private STNode parseVarDeclWithFunctionType(STNode typeDesc, STNode qualifiers, STNode metadata,
-                                                boolean isObjectMember, boolean isObjectTypeDesc) {
+    private STNode parseVarDeclWithFunctionType(STNode functionKeyword, STNode funcSignature, STNode qualifiers,
+                                                STNode metadata, boolean isObjectMember, boolean isObjectTypeDesc) {
+        endContext(); // end the func-type
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1012]: Can end the context in parseReturnTypeDescRhs and come here.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 5]
```diff
@@ -2222,9 +2233,11 @@ protected ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLo
                 return ParserRuleContext.TOP_LEVEL_NODE;
             case FUNC_DEF:
             case FUNC_DEF_OR_FUNC_TYPE:
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2235]: These two also can start with `isolated` keyword?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: moved to parseFunctionQualifiers method

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1037]: I c thax ayye will change.

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**